lang-js/src/main/java/io/crate/operation/language/JavaScriptUserDefinedFunction.java

```diff
@@ -0,0 +1,179 @@
+ * Agreement with Crate.io.
+ */
+
+package io.crate.operation.language;
+
+import io.crate.analyze.symbol.Symbol;
+import io.crate.data.Input;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.Scalar;
+import io.crate.types.*;
+import jdk.nashorn.api.scripting.ScriptObjectMirror;
+import jdk.nashorn.internal.runtime.ECMAException;
+import jdk.nashorn.internal.runtime.Undefined;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.lucene.BytesRefs;
+
+import javax.script.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class JavaScriptUserDefinedFunction extends Scalar<Object, Object> {
+
+    private final FunctionInfo info;
+    private final String script;
+
+    JavaScriptUserDefinedFunction(FunctionInfo info, String script) {
+        this.info = info;
+        this.script = script;
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+    @Override
+    public Scalar<Object, Object> compile(List<Symbol> arguments) {
+        try {
+            return new CompiledFunction(JavaScriptLanguage.bindScript(script));
+        } catch (ScriptException e) {
+            // this should not happen if the script was evaluated upfront
+            throw new UnsupportedOperationException("Could not evaluate JavaScript function", e);
+        }
+    }
+
+    @Override
+    public Object evaluate(Input<Object>[] values) {
+        try {
+            return evaluateScriptWithBindings(JavaScriptLanguage.bindScript(script), values);
+        } catch (ScriptException e) {
+            // this should not happen if the script was evaluated upfront
+            throw new UnsupportedOperationException("Could not evaluate JavaScript function", e);
+        }
+    }
+
+    private class CompiledFunction extends Scalar<Object, Object> {
+
+        private final Bindings bindings;
+
+        private CompiledFunction(Bindings bindings) {
+            this.bindings = bindings;
+        }
+
+        @Override
+        public FunctionInfo info() {
+            // return the functionInfo of the outer class,
+            // because the function info is the same for every compiled instance of a function
+            return info;
+        }
+
+        @Override
+        public final Object evaluate(Input<Object>[] values) {
+            return evaluateScriptWithBindings(bindings, values);
+        }
+
+    }
+
+    private Object evaluateScriptWithBindings(Bindings bindings, Input<Object>[] values) {
+        Object[] args = new Object[values.length];
+        for (int i = 0; i < values.length; i++) {
+            args[i] = processBytesRefInputIfNeeded(values[i].value());
+        }
+
+        Object result;
+        try {
+            result = ((ScriptObjectMirror) bindings.get(info.ident().name())).call(this, args);
+        } catch (NullPointerException e) {
+            throw new UnsupportedOperationException(
```

Comment:
pls check all UnsupportedOperationException usages, this is also wrong. probably illegalargument?